STYLEGUIDE.md

```diff
@@ -0,0 +1,215 @@
+```commandline
+source deactivate
+```
+
+### View list of exisiting virtual envs
+```commandline
+conda info --envs
+```
+
+## <a name="imports">Imports</a>
+------
+Refrain importing entire modules when only 1 object from the module is needed.
+
+**Don't:**
+```python
+import boto3, json, logging
+import pcf.core.PCF_exceptions 
+```
+
+**Do:**
+```python
+import boto3
+import json
+import logging
+
+from pcf.core.pcf_exceptions import NoResourceException
+```
+
+## <a name="naming">Naming</a>
+------
+PCF typically follows the same naming conventions as <a href="https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html">PEP8 standards</a>.
+
+### <a name="variables">Variables</a>  
+```python
+# Lowercase multi-word separated by an underscore
+my_greeting = "hello"
+```
+
+```python
+# Non-public instance variables should begin with a single underscore
+_client = ""
+```
+
+### <a name="classes">Classes</a>
+```python
+# UpperCaseCamelCase convention
+class LambaFunction():
+    # class
+
+class AWSResource():
+    # class
+
+class EC2Instance():
+    # class
+```
+
+### <a name="funcs">Fucntions/Methods</a>
+```python
+# Lowercase multi-word separated by an underscore
+def get_status(self):
+    pass
+```
+```python
+# Non-public methods should begin with a single underscore
+def _terminate(self):
+    pass
+```
+
+## <a name="spacing">Spacing</a>
+------
+* 4 spaces = 1 indent
+* Leave a single blank line after function/method declarations to aid in visual clarity and organization
+
+## <a name="strings">Strings</a>
+------
+### <a name="quotes">Quotes</a>  
+
+    In Python single quotes and double quotes are used interchangablely. We will stick to double quotes for consistency.  
+
+**Don't:**  
+```python
+nameOfSchool = 'Marshall'
+nameOfOffice = "Clarendon"
+```  
+
+**Do:**  
+```python
+nameOfSchool = "Marshall"
+nameOfOffice = "Clarendon"
+```   
+
+### <a name="docstr">Docstrings</a>  
+Docstrings act as documentation for method definitions of a class within PCF. Always use """triple double quotes""" around docstrings. For multi-line docstrings, place the closing qoutes on a line by itself.
+
+The docstrings should give a brief description of what the method/class is doing, explain arguments and their type, if any, and list what the method/class returns.
+
+For more info see the <a href="https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring">docs</a>.
+
+**Example:**
+```python
+def is_state_equivalent(self, state1, state2):
+    """
+    Determines if states are equivalent. Uses equivalent_states defined in the Glacier class.
+
+    Args:
+        state1 (State):
+        state1 (State):
+
+    Returns:
+        bool
+    """
+```
+
+## <a name="dict">Dictionary Key Value Retrieval</a>
+------
+**Don't:**
+```python
+self.bucket_name = self.desired_state_definition["Bucket"]
+```
+**Do:**
+```python
+# dict.get("Key", default=None)
+bucket_name = desired_state_definition.get("Bucket")
+```
+
```

Comment:
Somewhere, I might like to see something about returning empty dictionaries rather than `None` objects, for the sake of easy "if empty" tests, and general neatness.

Reply:
Sorry, I was unclear here. I want to talk about methods we write, not so much key-retrieval. If we are writing any methods that return dicts (a fair number of them) we should always return an empty dict rather than a None in the case that there is nothing to return, so that typing is consistent, and so that we can utilize the built in boolean valuation of dictionaries. I don't personally mind people returning None for a key that doesn't exist if that key is supposed to be a string or int. If the key retrieved is supposed to be another dictionary, this same rule should apply.